examples/step-85/step-85.cc

```diff
@@ -0,0 +1,565 @@
+          // we need a new class: FEImmersedSurfaceValues. In addition to the
+          // ordinary functions shape_value(..), shape_grad(..), etc., one can
+          // use its normal_vector(..)-function to get an outward normal to the
+          // immersed surface, $\Gamma$. In terms of the level set function,
+          // this normal reads
+          // @f{equation*}
+          //   n = \frac{\nabla \psi}{\| \nabla \psi \|}.
+          // @f}
+          const boost::optional<const FEImmersedSurfaceValues<dim> &>
+            surface_fe_values = non_matching_fe_values.get_surface_fe_values();
+
+          if (surface_fe_values)
+            for (const unsigned int q :
+                 surface_fe_values->quadrature_point_indices())
+              {
+                const Point<dim> &point =
+                  surface_fe_values->quadrature_point(q);
+                const Tensor<1, dim> &normal =
+                  surface_fe_values->normal_vector(q);
+                for (const unsigned int i : surface_fe_values->dof_indices())
+                  {
+                    for (const unsigned int j :
+                         surface_fe_values->dof_indices())
+                      {
+                        local_stiffness(i, j) +=
+                          (-normal * surface_fe_values->shape_grad(i, q) *
+                             surface_fe_values->shape_value(j, q) +
+                           -normal * surface_fe_values->shape_grad(j, q) *
+                             surface_fe_values->shape_value(i, q) +
+                           nitsche_parameter / cell_side_length *
+                             surface_fe_values->shape_value(i, q) *
+                             surface_fe_values->shape_value(j, q)) *
+                          surface_fe_values->JxW(q);
+                      }
+                    local_rhs(i) +=
+                      boundary_condition.value(point) *
+                      (nitsche_parameter / cell_side_length *
+                         surface_fe_values->shape_value(i, q) -
+                       normal * surface_fe_values->shape_grad(i, q)) *
+                      surface_fe_values->JxW(q);
+                  }
+              }
+
+          std::vector<types::global_dof_index> local_dof_indices(
+            n_dofs_per_cell);
+          cell->get_dof_indices(local_dof_indices);
+
+          stiffness_matrix.add(local_dof_indices, local_stiffness);
+          rhs.add(local_dof_indices, local_rhs);
+
+          // The assembly of the ghost penalty term is straight forward. As we
+          // iterate over the local faces, we first check if each face should
+          // have ghost penalty. The actual assembly is simple using
+          // FEInterfaceValues. Assembling in this we will traverse each
+          // internal face in the mesh twice, so in order to get the penalty
+          // constant we expect, we multiply the penalty term with a factor 1/2.
```

Comment:
What about using the MeshWorker infrastructure here (as in step-74)? I am not an expert in that but if I remember correctly it allow you to specify that a face  is visited only once.